lib/stages/visTranspose.cpp

```diff
@@ -178,23 +209,41 @@ void visTranspose::main_thread() {
     // offset for copying into buffer
     uint32_t offset = 0;
 
+    // The dataset ID we read from the frame
+    dset_id_t ds_id;
+    // The dataset ID of the state without the time axis
+    dset_id_t base_ds_id;
+
     uint64_t frame_size = 0;
 
-    // Wait for a frame in the input buffer in order to get the dataset ID
-    if ((wait_for_full_frame(in_buf, unique_name.c_str(), 0)) == nullptr) {
-        return;
-    }
-    auto frame = visFrameView(in_buf, 0);
-    dset_id_t ds_id = frame.dataset_id;
-    auto future_ds_state = std::async(&visTranspose::get_dataset_state, this, ds_id);
-
-    if (!future_ds_state.get()) {
-        FATAL_ERROR("Set to not use dataset_broker and couldn't find ancestor of dataset {}. "
-                    "Make sure there is a stage upstream in the config, that adds the dataset "
-                    "states.\nExiting...",
-                    ds_id);
+    bool found_frame = false;
+    uint32_t first_ind = 0;
+    while (!found_frame) {
+        // Wait for a frame in the input buffer in order to get the dataset ID
+        if ((wait_for_full_frame(in_buf, unique_name.c_str(), first_ind)) == nullptr) {
+            return;
+        }
+        auto frame = visFrameView(in_buf, first_ind);
+        if (frame.fpga_seq_length == 0) {
+            INFO("Got empty frame ({:d}).", first_ind);
+            first_ind++;
+            continue;
+        }
+        found_frame = true;
+
+        ds_id = frame.dataset_id;
+        if (!get_dataset_state(ds_id)) {
+            ERROR("Couldn't find ancestor of dataset {}. "
+                  "Make sure there is a stage upstream in the config, that adds the dataset "
+                  "states.\nExiting...",
+                  ds_id);
+            return;
+        }
     }
 
+    // Get the original dataset ID (before adding time axis)
+    base_ds_id = base_dset(ds_id);
```

Comment:
This is a bit of a stupid question asked much too late, but what's wrong about just using the datasetID that has the timeState instead of going back to the base dataset?